sklearn/neighbors/lmnn.py

```diff
@@ -0,0 +1,1013 @@
+
+
+    Notes
+    -----
+    Large margin nearest neighbor (LMNN) is a machine learning algorithm for
+    metric learning. It learns a (pseudo-)metric in a supervised fashion to
+    improve the classification accuracy of the k-nearest neighbor rule.
+    The main intuition behind LMNN is to learn a pseudometric under which all
+    data instances in the training set are surrounded by at least k instances
+    that share the same class label. If this is achieved, the leave-one-out
+    error is minimized.
+    This implementation follows closely Kilian Weinberger's MATLAB code found
+    at <https://bitbucket.org/mlcircus/lmnn> which solves the unconstrained
+    problem, finding a linear transformation with L-BFGS instead of solving the
+    constrained problem that finds the globally optimal metric. Different from
+    the paper, the problem solved by this implementation is with the squared
+    hinge loss (to make the problem differentiable).
+
+    .. warning::
+
+        At least for 32bit systems, one cannot expect precise reproducibility
+        of PCA and therefore the transformations in 2 identical runs can
+        diverge even before the first iteration of LargeMarginNearestNeighbor.
+        Therefore, one should not expect any reproducibility of the
+        transformations found by `LargeMarginNearestNeighbor` when
+        initializing with PCA (`init`='pca').
+
+
+    References
+    ----------
+    .. [1] Weinberger, Kilian Q., and Lawrence K. Saul. "Distance Metric
+    Learning for Large Margin Nearest Neighbor Classification."
+    Journal of Machine Learning Research, Vol. 10, Feb. 2009, pp. 207-244.
+    (http://jmlr.csail.mit.edu/papers/volume10/weinberger09a/weinberger09a.pdf)
+
+    .. [2] Wikipedia entry on Large Margin Nearest Neighbor
+    (https://en.wikipedia.org/wiki/Large_margin_nearest_neighbor)
+
+    """
+
+    def __init__(self, n_features_out=None, init='pca', warm_start=False,
+                 n_neighbors=3, algorithm='auto', targets=None,
+                 max_constraints=500000, use_sparse=True, max_iter=50,
+                 tol=1e-5, max_corrections=100, callback=None, verbose=0,
+                 random_state=None, n_jobs=1):
+
+        # Parameters
+        self.n_features_out = n_features_out
+        self.init = init
+        self.warm_start = warm_start
+        self.n_neighbors = n_neighbors
+        self.algorithm = algorithm
+        self.targets = targets
+        self.max_constraints = max_constraints
+        self.use_sparse = use_sparse
+        self.max_iter = max_iter
+        self.tol = tol
+        self.max_corrections = max_corrections
+        self.callback = callback
+        self.verbose = verbose
+        self.random_state = random_state
+        self.n_jobs = n_jobs
+
+    def fit(self, X, y):
+        """Find a linear transformation by optimization of the unconstrained
+        problem, such that the k-nearest neighbor classification accuracy
+        improves.
```

Comment:
does not follow pep257 rules with first 1 short headline and then a longer text.